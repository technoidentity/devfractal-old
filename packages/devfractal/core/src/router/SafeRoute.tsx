import React from 'react'
import { Route, RouteProps } from 'react-router-dom'
import { useSafeRouter } from './RouterContext'

export function SafeRoute({
  component: Component,
  render,
  children,
  exact,
  ...props
}: RouteProps): JSX.Element {
  const { setRouteMatched } = useSafeRouter()

  return (
    <Route
      {...props}
      exact={exact !== undefined ? exact : true}
      render={renderProps => {
        if (setRouteMatched) {
          setRouteMatched(true)
        }

        return Component ? (
          <Component {...renderProps} />
        ) : render ? (
          render(renderProps)
        ) : (
          children
        )
      }}
    />
  )
}
