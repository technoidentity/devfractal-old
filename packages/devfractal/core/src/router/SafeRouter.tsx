import React from 'react'
import {
  BrowserRouter,
  BrowserRouterProps,
  HashRouter,
  HashRouterProps,
  Route,
} from 'react-router-dom'
import { SafeRouterContext, useSafeRouter } from './RouterContext'

// tslint:disable no-class readonly-keyword no-object-mutation no-this
class RouteMatched {
  private routeMatched: boolean = false

  setRouteMatched = (routeMatched: boolean) => {
    this.routeMatched = routeMatched
  }

  getRouteMatched = () => this.routeMatched
}
// tslint:enable no-class readonly-keyword no-object-mutation no-this

export type RouterType = 'browser' | 'hash'

export type RouterProps<T extends RouterType> = (T extends 'browser'
  ? BrowserRouterProps
  : HashRouterProps) & {
  readonly variant?: T
  readonly children: React.ReactNode
}

const CheckRouteMatched: React.FC = () => {
  const { getRouteMatched } = useSafeRouter()
  if (!getRouteMatched()) {
    throw new Error('no route matched')
  }
  // tslint:disable-next-line: no-null-keyword
  return null
}

const RouterChildren: React.FC = ({ children }) => (
  <Route>
    {routeProps => (
      <SafeRouterContext.Provider
        value={{ ...routeProps, ...new RouteMatched() }}
      >
        {children}
        <CheckRouteMatched />
      </SafeRouterContext.Provider>
    )}
  </Route>
)

export function SafeRouter<T extends RouterType>({
  variant,
  children,
  ...props
}: RouterProps<T>): JSX.Element {
  return variant === 'browser' ? (
    <BrowserRouter {...props}>
      <RouterChildren>{children}</RouterChildren>
    </BrowserRouter>
  ) : (
    <HashRouter {...props}>
      <RouterChildren>{children}</RouterChildren>
    </HashRouter>
  )
}
