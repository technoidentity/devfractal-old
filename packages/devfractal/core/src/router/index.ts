export * from './SafeRouter'
export * from './RouterContext'
export * from './SafeRoute'
export * from './SimpleRedirect'
export * from './useParamsSafe'
