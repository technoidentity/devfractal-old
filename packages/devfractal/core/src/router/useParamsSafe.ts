import { useParams } from 'react-router'
import { cast, ObjC, ObjTypeOf, Props } from 'technoidentity-utils'

export function useParamsSafe<Opt extends Props, Req extends Props>(
  paramsSpec: ObjC<Opt, Req>,
): ObjTypeOf<Opt, Req> {
  // tslint:disable-next-line: typedef
  const params = useParams()
  return cast(paramsSpec, params)
}
