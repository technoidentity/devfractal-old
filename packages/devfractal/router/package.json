{
  "name": "@stp/router",
  "version": "1.2.19",
  "source": "src/index.ts",
  "main": "dist/index.js",
  "module": "dist/index.esm.js",
  "typings": "dist/index.d.ts",
  "scripts": {
    "build": "tsdx build --tsconfig tsconfig.build.json",
    "start": "parcel index.html --open",
    "lib": "tsdx watch --tsconfig tsconfig.build.json",
    "test": "jest --passWithNoTests",
    "test:watch": "yarn test --watch",
    "elint:fix": "yarn elint --quiet --fix",
    "elint": "eslint './src/**/*.{js,ts,tsx}'",
    "tlint": "yarn tslint -p . -c tslint.json 'src/**/*.tsx' 'src/**/*.ts'",
    "tlint:fix": "yarn lint --fix",
    "lint": "yarn elint && yarn tlint",
    "lint:fix": "yarn elint:fix && yarn tlint:fix",
    "types": "tsc --noEmit -p .",
    "types:watch": "cross-env TSC_WATCHFILE=UseFsEventsWithFallbackDynamicPolling yarn types --watch",
    "types:watch:verbose": "cross-env TSC_WATCHFILE=UseFsEventsWithFallbackDynamicPolling yarn types --watch --preserveWatchOutput",
    "sg:build": "styleguidist build",
    "sg": "styleguidist server",
    "validate": "yarn lint:fix && yarn types && yarn test",
    "validate:noFix": "yarn lint && yarn types && yarn test"
  },
  "dependencies": {
    "query-string": "6.11.0",
    "react-router": "5.1.2",
    "react-router-dom": "5.1.2",
    "tslib": "1.10.0"
  },
  "devDependencies": {
    "@stp/utils": "1.2.19",
    "cross-env": "7.0.0",
    "eslint": "6.8.0",
    "jest": "25.1.0",
    "parcel-bundler": "1.12.4",
    "prettier": "1.19.1",
    "tsdx": "0.12.3",
    "tslint": "6.0.0",
    "typescript": "3.7.5"
  },
  "peerDependencies": {
    "@stp/utils": "1.2.19",
    "react-dom": "16.12.0",
    "react": "16.12.0"
  }
}
