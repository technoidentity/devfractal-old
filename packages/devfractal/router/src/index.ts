export { useHistory, useLocation } from 'react-router'
export { Link, LinkProps, NavLink, Switch } from 'react-router-dom'
export * from './useParams'
export * from './Router'
export * from './RouterContext'
export * from './SafeRoute'
export * from './SimpleRedirect'
