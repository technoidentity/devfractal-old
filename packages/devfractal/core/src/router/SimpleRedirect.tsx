import React from 'react'
import { Redirect, RouteComponentProps } from 'react-router-dom'
import { useSafeRouter } from './RouterContext'
import { SafeRoute } from './SafeRoute'

export interface SimpleRedirectProps {
  readonly from: string
  readonly to: string
  readonly exact?: boolean
}

export const SimpleRedirect: React.FC<SimpleRedirectProps> = ({
  from,
  to,
  exact,
}) => {
  const { setRouteMatched } = useSafeRouter()
  return (
    <SafeRoute
      exact={exact !== undefined ? exact : true}
      path={from}
      render={() => {
        if (setRouteMatched) {
          setRouteMatched(true)
        }
        return <Redirect to={to} />
      }}
    />
  )
}

export type RouteComponentPropsRemoved<T> = Omit<T, keyof RouteComponentProps>

export function removeRouteComponentProps<T extends RouteComponentProps>(
  props: T,
): RouteComponentPropsRemoved<T> {
  const { match, location, history, staticContext, ...result } = props
  return result
}
