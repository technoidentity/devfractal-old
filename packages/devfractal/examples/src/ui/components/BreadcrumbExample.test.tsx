import { Router } from '@stp/router'
import React from 'react'
import renderer, { ReactTestRenderer } from 'react-test-renderer'
import { BreadcrumbExample } from './BreadcrumbExample'

const Ex: React.FC = () => (
  <Router variant="hash">
    <BreadcrumbExample />
  </Router>
)

test.skip('Breadcrumb Example', () => {
  const component: ReactTestRenderer = renderer.create(<Ex />)
  expect(component.toJSON()).toMatchSnapshot()
})
