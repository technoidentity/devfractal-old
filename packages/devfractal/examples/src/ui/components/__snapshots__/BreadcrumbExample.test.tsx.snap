// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb Example 1`] = `
Array [
  <section
    className="section"
  >
    <nav
      className="breadcrumb"
    >
      <ul>
        <li
          className=""
          value="alignments"
        >
          <a
            aria-current={null}
            href="#/components/breadcrumb/alignments"
            onClick={[Function]}
          >
            Alignment
          </a>
        </li>
        <li
          className=""
          value="icons"
        >
          <a
            aria-current={null}
            href="#/components/breadcrumb/icons"
            onClick={[Function]}
          >
            Icons
          </a>
        </li>
        <li
          className=""
          value="separators"
        >
          <a
            aria-current={null}
            href="#/components/breadcrumb/separators"
            onClick={[Function]}
          >
            Separators
          </a>
        </li>
        <li
          className=""
          value="sizes"
        >
          <a
            aria-current={null}
            href="#/components/breadcrumb/sizes"
            onClick={[Function]}
          >
            Sizes
          </a>
        </li>
      </ul>
    </nav>
  </section>,
  <div
    className="container"
  />,
]
`;
